src/torrent.cpp

```diff
@@ -5399,8 +5368,27 @@ namespace libtorrent
 		peers_erased(st.erased);
 
 		p->set_peer_info(nullptr);
-		TORRENT_ASSERT(i != m_connections.end());
-		m_connections.erase(i);
+
+		update_want_peers();
+		update_want_tick();
+	}
+
+	void torrent::on_remove_peers()
+	{
+		TORRENT_ASSERT(is_single_thread());
+		INVARIANT_CHECK;
+
+		std::vector<peer_connection*> peers;
+		m_peers_to_disconnect.swap(peers);
+		for (peer_connection* p : peers)
+		{
+			TORRENT_ASSERT(p != nullptr);
+
+			auto const i = sorted_find(m_connections, p);
```

Comment:
I wonder if having both sorted `m_connections` and `m_peers_to_disconnect` reduces the `O(n^2)`, or if it does't worth the effort.